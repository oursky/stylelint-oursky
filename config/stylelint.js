module.exports = {
  rules: {
    // NOTE(SCSS): sass does not compile percentage into number.
    "alpha-value-notation": "number",
    "at-rule-allowed-list": null,
    "at-rule-disallowed-list": null,
    "at-rule-empty-line-before": null,
    "at-rule-name-case": "lower",
    "at-rule-name-newline-after": null,
    "at-rule-name-space-after": null,
    "at-rule-no-unknown": true,
    "at-rule-no-vendor-prefix": true,
    "at-rule-property-required-list": null,
    "at-rule-semicolon-newline-after": null,
    "at-rule-semicolon-space-before": null,
    "block-closing-brace-empty-line-before": null,
    "block-closing-brace-newline-after": null,
    "block-closing-brace-newline-before": null,
    "block-closing-brace-space-after": null,
    "block-closing-brace-space-before": null,
    "block-no-empty": null,
    "block-opening-brace-newline-after": null,
    "block-opening-brace-newline-before": null,
    "block-opening-brace-space-after": null,
    "block-opening-brace-space-before": null,
    // NOTE(SCSS): sass compiles color function into hex.
    "color-function-notation": "legacy",
    "color-hex-alpha": null,
    "color-hex-case": null,
    "color-hex-length": null,
    "color-named": null,
    "color-no-hex": null,
    "color-no-invalid-hex": true,
    "comment-empty-line-before": null,
    "comment-no-empty": null,
    "comment-pattern": null,
    "comment-whitespace-inside": null,
    "comment-word-disallowed-list": null,
    "custom-media-pattern": null,
    "custom-property-empty-line-before": null,
    "custom-property-pattern": null,
    "custom-property-no-missing-var-function": true,
    "declaration-bang-space-after": null,
    "declaration-bang-space-before": null,

    "declaration-block-no-duplicate-properties": [
      true,
      {
        ignore: ["consecutive-duplicates-with-different-values"],
      },
    ],
    "declaration-block-no-duplicate-custom-properties": true,

    "declaration-block-no-redundant-longhand-properties": [
      true,
      {
        ignoreShorthands: ["/flex/"],
      },
    ],

    "declaration-block-no-shorthand-property-overrides": true,
    "declaration-block-semicolon-newline-after": null,
    "declaration-block-semicolon-newline-before": null,
    "declaration-block-semicolon-space-after": null,
    "declaration-block-semicolon-space-before": null,
    "declaration-block-single-line-max-declarations": null,
    "declaration-block-trailing-semicolon": null,
    "declaration-colon-newline-after": null,
    "declaration-colon-space-after": null,
    "declaration-colon-space-before": null,
    "declaration-empty-line-before": null,
    "declaration-no-important": true,
    "declaration-property-max-values": null,
    "declaration-property-unit-allowed-list": null,
    "declaration-property-unit-disallowed-list": null,
    "declaration-property-value-allowed-list": null,
    "declaration-property-value-disallowed-list": null,
    "font-family-name-quotes": "always-unless-keyword",
    "font-family-no-duplicate-names": true,
    "font-family-no-missing-generic-family-keyword": true,

    "font-weight-notation": [
      "numeric",
      {
        ignore: ["relative"],
      },
    ],

    "function-allowed-list": null,
    "function-disallowed-list": null,
    "function-calc-no-unspaced-operator": true,
    "function-comma-newline-after": null,
    "function-comma-newline-before": null,
    "function-comma-space-after": null,
    "function-comma-space-before": null,
    "function-linear-gradient-no-nonstandard-direction": true,
    "function-max-empty-lines": null,
    "function-name-case": "lower",
    "function-no-unknown": true,
    "function-parentheses-newline-inside": null,
    "function-parentheses-space-inside": null,
    "function-url-no-scheme-relative": null,
    "function-url-quotes": "always",
    "function-url-scheme-allowed-list": null,
    "function-url-scheme-disallowed-list": null,
    "function-whitespace-after": null,
    "hue-degree-notation": "angle",
    indentation: null,
    "keyframe-declaration-no-important": true,
    "keyframes-name-pattern": null,
    "length-zero-no-unit": true,
    linebreaks: null,
    "max-empty-lines": null,
    "max-line-length": null,
    "max-nesting-depth": null,
    "media-feature-colon-space-after": null,
    "media-feature-colon-space-before": null,
    "media-feature-name-allowed-list": null,
    "media-feature-name-disallowed-list": null,
    "media-feature-name-case": "lower",
    "media-feature-name-no-unknown": true,
    "media-feature-name-no-vendor-prefix": true,
    "media-feature-name-value-allowed-list": null,
    "media-feature-parentheses-space-inside": null,
    "media-feature-range-operator-space-after": null,
    "media-feature-range-operator-space-before": null,
    "media-query-list-comma-newline-after": null,
    "media-query-list-comma-newline-before": null,
    "media-query-list-comma-space-after": null,
    "media-query-list-comma-space-before": null,
    "no-descending-specificity": true,
    "no-duplicate-at-import-rules": true,
    "no-duplicate-selectors": true,
    "no-empty-first-line": null,
    "no-empty-source": null,
    "no-eol-whitespace": null,
    "no-extra-semicolons": null,
    // NOTE(SCSS): sass has line comments
    "no-invalid-double-slash-comments": true,
    "no-invalid-position-at-import-rule": true,
    "no-irregular-whitespace": null,
    "no-missing-end-of-source-newline": null,
    "no-unknown-animations": true,
    "named-grid-areas-no-invalid": true,
    "number-leading-zero": "always",
    "number-max-precision": 2,
    "number-no-trailing-zeros": true,
    "property-allowed-list": null,
    "property-disallowed-list": null,
    "property-case": "lower",
    "property-no-unknown": true,
    "property-no-vendor-prefix": true,
    "rule-empty-line-before": null,
    "rule-selector-property-disallowed-list": null,
    "selector-attribute-brackets-space-inside": null,
    "selector-attribute-name-disallowed-list": null,
    "selector-attribute-operator-allowed-list": null,
    "selector-attribute-operator-disallowed-list": null,
    "selector-attribute-operator-space-after": null,
    "selector-attribute-operator-space-before": null,
    "selector-attribute-quotes": "always",
    "selector-class-pattern": null,
    "selector-combinator-allowed-list": null,
    "selector-combinator-disallowed-list": null,
    "selector-combinator-space-after": null,
    "selector-combinator-space-before": null,
    "selector-disallowed-list": null,
    "selector-descendant-combinator-no-non-space": true,
    "selector-id-pattern": null,
    "selector-list-comma-newline-after": null,
    "selector-list-comma-newline-before": null,
    "selector-list-comma-space-after": null,
    "selector-list-comma-space-before": null,
    "selector-max-attribute": null,
    "selector-max-class": null,
    "selector-max-combinators": null,
    "selector-max-compound-selectors": null,
    "selector-max-empty-lines": null,
    "selector-max-id": null,
    "selector-max-pseudo-class": null,
    "selector-max-specificity": null,
    "selector-max-type": null,
    "selector-max-universal": null,
    "selector-nested-pattern": null,
    "selector-no-qualifying-type": null,
    "selector-no-vendor-prefix": true,
    "selector-pseudo-class-allowed-list": null,
    "selector-pseudo-class-disallowed-list": null,
    "selector-pseudo-class-case": "lower",
    "selector-pseudo-class-no-unknown": true,
    "selector-pseudo-class-parentheses-space-inside": null,
    "selector-pseudo-element-allowed-list": null,
    "selector-pseudo-element-disallowed-list": null,
    "selector-pseudo-element-case": "lower",
    "selector-pseudo-element-colon-notation": "double",
    "selector-pseudo-element-no-unknown": true,
    "selector-type-case": "lower",
    "selector-type-no-unknown": true,
    "shorthand-property-no-redundant-values": true,
    "string-no-newline": true,
    "string-quotes": null,
    "time-min-milliseconds": null,
    "unicode-bom": null,
    "unit-allowed-list": null,
    "unit-disallowed-list": null,
    "unit-case": "lower",
    "unit-no-unknown": true,
    "value-keyword-case": "lower",
    "value-list-comma-newline-after": null,
    "value-list-comma-newline-before": null,
    "value-list-comma-space-after": null,
    "value-list-comma-space-before": null,
    "value-list-max-empty-lines": null,
    "value-no-vendor-prefix": true,
  },
};
