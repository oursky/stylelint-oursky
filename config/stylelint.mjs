export default {
  rules: {
    // NOTE(SCSS): sass does not compile percentage into number.
    "alpha-value-notation": "number",
    "annotation-no-unknown": true,
    "at-rule-allowed-list": null,
    "at-rule-disallowed-list": null,
    "at-rule-empty-line-before": null,
    "at-rule-no-unknown": true,
    "at-rule-no-vendor-prefix": true,
    "at-rule-property-required-list": null,
    "block-no-empty": null,
    // NOTE(SCSS): sass compiles color function into hex.
    "color-function-notation": "legacy",
    "color-hex-alpha": null,
    "color-hex-length": null,
    "color-named": null,
    "color-no-hex": null,
    "color-no-invalid-hex": true,
    "comment-empty-line-before": null,
    "comment-no-empty": null,
    "comment-pattern": null,
    "comment-whitespace-inside": null,
    "comment-word-disallowed-list": null,
    "custom-media-pattern": null,
    "custom-property-empty-line-before": null,
    "custom-property-pattern": null,
    "custom-property-no-missing-var-function": true,

    "declaration-block-no-duplicate-properties": [
      true,
      {
        ignore: ["consecutive-duplicates-with-different-values"],
      },
    ],
    "declaration-block-no-duplicate-custom-properties": true,

    "declaration-block-no-redundant-longhand-properties": null,

    "declaration-block-no-shorthand-property-overrides": true,
    "declaration-block-single-line-max-declarations": null,
    "declaration-empty-line-before": null,
    "declaration-no-important": true,
    "declaration-property-max-values": null,
    "declaration-property-unit-allowed-list": null,
    "declaration-property-unit-disallowed-list": null,
    "declaration-property-value-allowed-list": null,
    "declaration-property-value-disallowed-list": null,
    "declaration-property-value-no-unknown": null,
    "font-family-name-quotes": "always-unless-keyword",
    "font-family-no-duplicate-names": true,
    "font-family-no-missing-generic-family-keyword": true,

    "font-weight-notation": [
      "numeric",
      {
        ignore: ["relative"],
      },
    ],

    "function-allowed-list": null,
    "function-disallowed-list": null,
    "function-calc-no-unspaced-operator": true,
    "function-linear-gradient-no-nonstandard-direction": true,
    "function-name-case": "lower",
    "function-no-unknown": true,
    "function-url-no-scheme-relative": null,
    "function-url-quotes": "always",
    "function-url-scheme-allowed-list": null,
    "function-url-scheme-disallowed-list": null,
    "hue-degree-notation": "angle",
    "import-notation": null,
    "keyframe-block-no-duplicate-selectors": true,
    "keyframe-declaration-no-important": true,
    "keyframe-selector-notation": null,
    "keyframes-name-pattern": null,
    "length-zero-no-unit": true,
    "lightness-notation": null,
    "max-nesting-depth": null,
    "media-feature-name-allowed-list": null,
    "media-feature-name-disallowed-list": null,
    "media-feature-name-no-unknown": true,
    "media-feature-name-no-vendor-prefix": true,
    "media-feature-name-value-allowed-list": null,
    "media-feature-name-value-no-unknown": null,
    "media-feature-name-unit-allowed-list": null,
    "media-feature-range-notation": null,
    "media-query-no-invalid": true,
    "no-descending-specificity": true,
    "no-duplicate-at-import-rules": true,
    "no-duplicate-selectors": true,
    "no-empty-source": null,
    // NOTE(SCSS): sass has line comments
    "no-invalid-double-slash-comments": true,
    "no-invalid-position-at-import-rule": true,
    "no-irregular-whitespace": null,
    "no-unknown-animations": true,
    "no-unknown-custom-media": null,
    "no-unknown-custom-properties": null,
    "named-grid-areas-no-invalid": true,
    "number-max-precision": null,
    "property-allowed-list": null,
    "property-disallowed-list": null,
    "property-no-unknown": true,
    "property-no-vendor-prefix": true,
    "rule-empty-line-before": null,
    "rule-selector-property-disallowed-list": null,
    "selector-anb-no-unmatchable": true,
    "selector-attribute-name-disallowed-list": null,
    "selector-attribute-operator-allowed-list": null,
    "selector-attribute-operator-disallowed-list": null,
    "selector-attribute-quotes": "always",
    "selector-class-pattern": null,
    "selector-combinator-allowed-list": null,
    "selector-combinator-disallowed-list": null,
    "selector-disallowed-list": null,
    "selector-id-pattern": null,
    "selector-max-attribute": null,
    "selector-max-class": null,
    "selector-max-combinators": null,
    "selector-max-compound-selectors": null,
    "selector-max-id": null,
    "selector-max-pseudo-class": null,
    "selector-max-specificity": null,
    "selector-max-type": null,
    "selector-max-universal": null,
    "selector-nested-pattern": null,
    "selector-no-qualifying-type": null,
    "selector-no-vendor-prefix": true,
    "selector-not-notation": null,
    "selector-pseudo-class-allowed-list": null,
    "selector-pseudo-class-disallowed-list": null,
    "selector-pseudo-class-no-unknown": true,
    "selector-pseudo-element-allowed-list": null,
    "selector-pseudo-element-disallowed-list": null,
    "selector-pseudo-element-colon-notation": "double",
    "selector-pseudo-element-no-unknown": true,
    "selector-type-case": "lower",
    "selector-type-no-unknown": true,
    "shorthand-property-no-redundant-values": true,
    "string-no-newline": true,
    "time-min-milliseconds": null,
    "unit-allowed-list": null,
    "unit-disallowed-list": null,
    "unit-no-unknown": true,
    "value-keyword-case": "lower",
    "value-no-vendor-prefix": true,
  },
};
