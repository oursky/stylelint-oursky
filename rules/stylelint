alpha-value-notation
annotation-no-unknown
at-rule-allowed-list
at-rule-disallowed-list
at-rule-empty-line-before
at-rule-no-unknown
at-rule-no-vendor-prefix
at-rule-property-required-list
block-no-empty
color-function-notation
color-hex-alpha
color-hex-length
color-named
color-no-hex
color-no-invalid-hex
comment-empty-line-before
comment-no-empty
comment-pattern
comment-whitespace-inside
comment-word-disallowed-list
custom-media-pattern
custom-property-empty-line-before
custom-property-no-missing-var-function
custom-property-pattern
declaration-block-no-duplicate-custom-properties
declaration-block-no-duplicate-properties
declaration-block-no-redundant-longhand-properties
declaration-block-no-shorthand-property-overrides
declaration-block-single-line-max-declarations
declaration-empty-line-before
declaration-no-important
declaration-property-max-values
declaration-property-unit-allowed-list
declaration-property-unit-disallowed-list
declaration-property-value-allowed-list
declaration-property-value-disallowed-list
declaration-property-value-no-unknown
font-family-name-quotes
font-family-no-duplicate-names
font-family-no-missing-generic-family-keyword
font-weight-notation
function-allowed-list
function-calc-no-unspaced-operator
function-disallowed-list
function-linear-gradient-no-nonstandard-direction
function-name-case
function-no-unknown
function-url-no-scheme-relative
function-url-quotes
function-url-scheme-allowed-list
function-url-scheme-disallowed-list
hue-degree-notation
import-notation
keyframe-block-no-duplicate-selectors
keyframe-declaration-no-important
keyframe-selector-notation
keyframes-name-pattern
length-zero-no-unit
lightness-notation
max-nesting-depth
media-feature-name-allowed-list
media-feature-name-disallowed-list
media-feature-name-no-unknown
media-feature-name-no-vendor-prefix
media-feature-name-unit-allowed-list
media-feature-name-value-allowed-list
media-feature-name-value-no-unknown
media-feature-range-notation
media-query-no-invalid
named-grid-areas-no-invalid
no-descending-specificity
no-duplicate-at-import-rules
no-duplicate-selectors
no-empty-source
no-invalid-double-slash-comments
no-invalid-position-at-import-rule
no-irregular-whitespace
no-unknown-animations
no-unknown-custom-media
no-unknown-custom-properties
number-max-precision
property-allowed-list
property-disallowed-list
property-no-unknown
property-no-vendor-prefix
rule-empty-line-before
rule-selector-property-disallowed-list
selector-anb-no-unmatchable
selector-attribute-name-disallowed-list
selector-attribute-operator-allowed-list
selector-attribute-operator-disallowed-list
selector-attribute-quotes
selector-class-pattern
selector-combinator-allowed-list
selector-combinator-disallowed-list
selector-disallowed-list
selector-id-pattern
selector-max-attribute
selector-max-class
selector-max-combinators
selector-max-compound-selectors
selector-max-id
selector-max-pseudo-class
selector-max-specificity
selector-max-type
selector-max-universal
selector-nested-pattern
selector-no-qualifying-type
selector-no-vendor-prefix
selector-not-notation
selector-pseudo-class-allowed-list
selector-pseudo-class-disallowed-list
selector-pseudo-class-no-unknown
selector-pseudo-element-allowed-list
selector-pseudo-element-colon-notation
selector-pseudo-element-disallowed-list
selector-pseudo-element-no-unknown
selector-type-case
selector-type-no-unknown
shorthand-property-no-redundant-values
string-no-newline
time-min-milliseconds
unit-allowed-list
unit-disallowed-list
unit-no-unknown
value-keyword-case
value-no-vendor-prefix
