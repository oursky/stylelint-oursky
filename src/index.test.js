import stylelint from "stylelint";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

async function lint(code) {
  const result = await stylelint.lint({
    code,
    formatter: "json",
    config: {
      plugins: [join(__dirname, "./index.mjs")],
      rules: {
        "oursky/flex": "error",
      },
    },
  });

  const report = JSON.parse(result.report);
  const reportWithoutSource = report.map((a) => {
    return {
      ...a,
      source: undefined,
    };
  });
  return JSON.stringify(reportWithoutSource);
}

async function check(code) {
  const output = await lint(code);
  expect(output).toMatchSnapshot();
}

it("warns 1-value forms", async () => {
  return check("a{flex: 1;}");
});

it("warns 2-value forms", async () => {
  return check("a{flex: 1 0;}");
});

it("handles spaces", async () => {
  return check("a{flex: 1   0    ;}");
});

it("warns unitless flex-basis", async () => {
  return check("a{flex: 1   0  0  ;}");
});

it("warns unitless flex-basis of any number", async () => {
  return check("a{flex: 1   0  1  ;}");
});

it("passes well-formed flex", async () => {
  return check("a{flex: 1   0  0px  ;}");
});

it("warns flex-grow", async () => {
  return check("a{flex-grow: 1;}");
});

it("warns flex-shrink", async () => {
  return check("a{flex-shrink: 1;}");
});

it("warns flex-basis", async () => {
  return check("a{flex-basis: 1;}");
});

it("does not warn any other flex-*", async () => {
  return check("a{flex-direction: row;}");
});
