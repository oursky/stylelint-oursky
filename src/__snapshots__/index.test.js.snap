// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not warn any other flex-* 1`] = `"[{"source":"<input css Q-vYhk>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":false,"warnings":[]}]"`;

exports[`handles spaces 1`] = `"[{"source":"<input css nU8XED>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"Expected 3-values form of flex: <flex-grow> <flex-shrink> <flex-basis>; (oursky/flex)"}]}]"`;

exports[`passes well-formed flex 1`] = `"[{"source":"<input css 6D4nmu>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":false,"warnings":[]}]"`;

exports[`warns 1-value forms 1`] = `"[{"source":"<input css uR9paM>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"Expected 3-values form of flex: <flex-grow> <flex-shrink> <flex-basis>; (oursky/flex)"}]}]"`;

exports[`warns 2-value forms 1`] = `"[{"source":"<input css GRBR9Q>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"Expected 3-values form of flex: <flex-grow> <flex-shrink> <flex-basis>; (oursky/flex)"}]}]"`;

exports[`warns flex-basis 1`] = `"[{"source":"<input css LaeHnF>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"flex-[grow|shrink|basis] are not supported on IE 10. Use flex. (oursky/flex)"}]}]"`;

exports[`warns flex-grow 1`] = `"[{"source":"<input css kVghNh>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"flex-[grow|shrink|basis] are not supported on IE 10. Use flex. (oursky/flex)"}]}]"`;

exports[`warns flex-shrink 1`] = `"[{"source":"<input css wJOyJN>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"flex-[grow|shrink|basis] are not supported on IE 10. Use flex. (oursky/flex)"}]}]"`;

exports[`warns unitless flex-basis 1`] = `"[{"source":"<input css yoyoLZ>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"flex-basis: 0 not allowed. Please include unit. (oursky/flex)"}]}]"`;

exports[`warns unitless flex-basis of any number 1`] = `"[{"source":"<input css txrIOi>","deprecations":[],"invalidOptionWarnings":[],"parseErrors":[],"errored":true,"warnings":[{"line":1,"column":3,"rule":"oursky/flex","severity":"error","text":"flex-basis: 1 not allowed. Please include unit. (oursky/flex)"}]}]"`;
